### DISCUSSION TITLE
is QAT still on the roadmap for keras3?

### DISCUSSION THREAD

**Discussion by matpalm** [keras-team/keras#20319] [Q&A] [answered]
is QAT still on the roadmap?

( https://github.com/keras-team/keras/issues/18930 from Feb has it as a "near future" release. )

have been porting code to keras3 ( as a way to move more to Jax ) & have options for post training quantisation but expect a non trivial benefit from QAT in a number of projects.

can see a path forward by partially porting pieces of https://www.tensorflow.org/model_optimization/api_docs/python/tfmot and/or https://github.com/google/aqt but will hold back if a QAT api is imminent ? ( additionally; might have some bandwidth to help if there are community contrib options? )

**Comment by matpalm** [comment]
*bump* 

though i can jump through hoops to get tfmot working ( ish ) the fact it is keras2 compat only is a pain :/

if keras3 QAT is pending i can hold out, otherwise will have to consider dropping back to keras2 :( :( :(

**Comment by fchollet** [comment]
QAT is still on the roadmap but the work hasn't started yet, as we've had to prioritize other things... we can definitely re-prioritize it though. Most of the infra is already here in via the work @james77777778 did for float8 training.

**Comment by matpalm** [comment]
great! if there's any pointers you can give me @james77777778 i'd be happy to help out; should have a bit of bandwidth for this over the next few months.

**Reply by james77777778** [reply, marked as answer]
I'm happy to help with this but unfortunately I don't have enough bandwidth right now.

@matpalm I suggest starting by implementing these ops in a backend-agnostic way, as they are necessary for QAT:
- `tf.quantization.fake_quant_with_min_max_args`
- `tf.quantization.fake_quant_with_min_max_vars`
- `tf.quantization.fake_quant_with_min_max_vars_per_channel`

You should be able to implement them using `ops.custom_gradient`

**Reply by matpalm** [reply]
great thanks, i'll follow up in a PR

**Reply by james77777778** [reply]
I can help review the implementation if that sounds good to @fchollet!

**Reply by fchollet** [reply]
Awesome, thanks you two!

**Comment by doncarlos999** [comment]
@james77777778 Now that #20641 is merged, what would the next step in porting QAT from tfmot?

**Reply by james77777778** [reply]
I’m working on some POCs to verify the impl.
Currently, I think we need a wrapper to add the fake_quant op to the `call` method of the layers.

Ref: https://github.com/tensorflow/model-optimization/blob/master/tensorflow_model_optimization/python/core/quantization/keras/quantize_wrapper.py#L43

**Reply by dbrandonk** [reply]
Hi, is the remainder of this on anyone's radar?
